### PR TITLE
Add oom detection test

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Test to identify if there are any OOM events in the cluster.

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
cc @smarterclayton @rphillips @mrunalp @wking

File: test/extended/node/oom_detection.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+			g.Fail(fmt.Sprintf("Unexpected error: %v", err))
+		}
+
+		oomEncounteredEvents := []string{}
+		errorMsg := "System OOM encountered"
+		for _, event := range events.Items {
+			if !strings.Contains(event.Message, errorMsg) {
+				continue
+			}
+			oomEncounteredEvent := fmt.Sprintf("OOM encountered within node %v", event.InvolvedObject.Name)
```
Comment (wking) [inline test/extended/node/oom_detection.go:34]: Do the events say what got OOMed?  That would make it easier to route to the team in charge of the greedy component (for cases where OOM picked the right greedy component).

File: test/extended/node/oom_detection.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+			oomEncounteredEvents = append(oomEncounteredEvents, oomEncounteredEvent)
+		}
+		failMessage := fmt.Sprintf("%d OOM events occured", oomEncounteredEvents)
```
Comment (wking) [inline test/extended/node/oom_detection.go:37]: The count is nice, but having a list of OOMed components here would make routing the resulting bugs easier.

File: test/extended/node/oom_detection.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:34]: I thought about it but the message only talks about the [victim process](https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/pkg/kubelet/oom/oom_watcher_linux.go#L79). Not sure, if it be beneficial

File: test/extended/node/oom_detection.go (hunk 2 above)
Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:37]: I think identifying the component which caused OOMKiller(the greedy one) invocation is benefical but AFAIK we can get that info from journal logs.

File: test/extended/node/oom_detection.go (hunk 1 above)
Comment (mrunalp) [reply test/extended/node/oom_detection.go:34]: The process that ends up being the cause of the  OOM Killer invocation isn't always at fault. Any memory allocation by any process could cause the oom killer to be invoked.

**Comment by nstielau** [comment]
I wouldn't block this, but almost tempted to, until we could also measure the same metric across the fleet in aggregate.  If we can solve this problem now, we can use that pattern to get hard-failures in CI and soft-failures in aggregate fleet metrics.  I think that would significantly improve the signal we get.

The two ways ths would be possible would be:
* Write an Alert for this, and check for the firing of the alert rather than the symptom itself
* Collect telemetry for this, and check against that data directly.

I'm sure there are drawbacks to both of these, but I think being able to go from a single test to a single cluster to a subset of clusters to the whole fleet looking at the same signal is powerful compared to one more test.

**Comment by ravisantoshgudimetla** [comment]
> Write an Alert for this, and check for the firing of the alert rather than the symptom itself

I think we already have an alert for these situations. Something like:

`
{alertname="KubeNodeNotReady",severity="warning"} 
`
I think the main problem here is` from the alert how to identify if it's a kernel bug or resource pressure induced by a workload` which is orthogonal to the problem however I agree that having this info captured in a metric and storing it might be useful.

File: test/extended/node/oom_detection.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	oc := exutil.NewCLIWithoutNamespace("no-oom-kills within worker node journal logs")
+
+	g.It("OOM Killer shouldn't have been invoked in kubelet logs", func() {
+
+		nodeList, err := oc.AdminKubeClient().CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		oomJournalNodeList := []string{}
+		for _, node := range nodeList.Items {
+			// Choosing 150 min as the maximum upgrade duration has been set to 150m in upgrade.go
+			out, err := oc.Run("adm", "node-logs").Args(node.Name, "--unit=kubelet",
```
Comment (mrunalp) [inline test/extended/node/oom_detection.go:58]: These would be in the kernel logs. So either `dmesg` or `journalctl -k` equivalent.

Comment (mrunalp) [reply test/extended/node/oom_detection.go:58]: Or get the full journal and grep should be able to find it.

Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:58]: I saw entries like in the kubelet log:

```
Jan 06 23:27:40.051244 ip-10-0-225-225 kernel: prometheus invoked oom-killer: gfp_mask=0x6200ca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=-997
```

Do we need more info that we need to get from kernel logs?

Comment (mrunalp) [reply test/extended/node/oom_detection.go:58]: That looks like a kernel log entry to me.

Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:58]: I see, let me remove the `--unit=kubelet` and use `--path=journal`, perhaps we can get entire journal log.


This is what help suggests.
```
--path='journal': Retrieve the specified path within the node's /var/logs/ folder. The 'journal' value will allow
querying the journal on supported operating systems.
```
So, I believe we should be good.

File: test/extended/node/oom_detection.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	g.It("should not have pods that are OOM Killed", func() {
+		kubeClient := oc.AdminKubeClient()
+
+		events, err := kubeClient.CoreV1().Events("").List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			g.Fail(fmt.Sprintf("Unexpected error: %v", err))
+		}
+		oomEventNodeList := []string{}
+		errorMsg := "System OOM encountered"
```
Comment (smarterclayton) [inline test/extended/node/oom_detection.go:29]: You need to filter by suite start (see other event types where we filter those out).  We will eventually be running multiple suites in order - we might run disruptive (which causes OOMs) before we run e2e-conformanec (to verify that everything works after ooms).  If you look back before the suite start you may cause failures.

File: test/extended/node/oom_detection.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	oc := exutil.NewCLIWithoutNamespace("no-oom-kills within worker node journal logs")
+
+	g.It("OOM Killer shouldn't have been invoked in kubelet logs", func() {
+
+		nodeList, err := oc.AdminKubeClient().CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		oomJournalNodeList := []string{}
+		for _, node := range nodeList.Items {
+			// Choosing 150 min as the maximum upgrade duration has been set to 150m in upgrade.go
+			out, err := oc.Run("adm", "node-logs").Args(node.Name, "--path=journal",
```
Comment (smarterclayton) [inline test/extended/node/oom_detection.go:58]: You can get the exact test duration from test util - `DurationSinceStartInSeconds()`

File: test/extended/node/oom_detection.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+				"--case-sensitive=false", "--grep=invoked oom-killer", "--since=-150m").Output()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if !strings.Contains(out, "invoked oom-killer") {
```
Comment (smarterclayton) [inline test/extended/node/oom_detection.go:61]: Were you finding lines returned by --grep that did contain invoked oom killer?  Also, you're doing a case insensitive search above, and then this is case sensitive - why?

File: test/extended/node/oom_detection.go (hunk 4 above)
Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:29]: Ok makes sense. Let me look at that.

File: test/extended/node/oom_detection.go (hunk 6 above)
Comment (ravisantoshgudimetla) [reply test/extended/node/oom_detection.go:61]: > Were you finding lines returned by --grep that did contain invoked oom killer?

If even it doesn't contain `invoked oom killer`, it was still returning output like:

```
-- Logs begin at Mon 2021-03-01 14:50:39 UTC, end at Mon 2021-03-01 17:10:35 UTC. --
-- No entries --
-- Logs begin at Mon 2021-03-01 14:50:39 UTC, end at Mon 2021-03-01 17:10:35 UTC. --
-- No entries --
```
So, I added the `strings.Contains` check, I think the case sensitivity can be ignored. Will remove it in the next commit.

**Pull request closed without merging**